src/conn.rs

```diff
@@ -249,9 +249,16 @@ impl<const N: usize, P: ConnectionPeer> Connection<N, P> {
         tokio::pin!(idle_timeout);
         loop {
             tokio::select! {
+                biased;
```

Comment:
Ah, that's an interesting approach that hadn't occurred to me.

The tradeoffs I'm seeing on that choice are:

Pros of switching to `try_recv`
- Fewer opportunities to mess up the branch order in subtle ways
- `biased` is a less-familiar approach, it requires more understanding from the contributors

Cons of switching to `try_recv`
- Slight performance loss (`biased` is a bit faster)
- In maintenance, we'd want to make sure that any path that resets `idle_timeout` is also double-checked in the timeout branch. For example here, I guess we'd also want to check the `writes.try_recv()`
- More lines of code: the timeout processing gets messier. Do we just restart the loop if there is a packet waiting on the queue (in which case we might loop through a random number of times before hitting the branch we want), or we could actually process the packet in the idle-timeout branch, which seems surprising and leaves room open for bugs where someone makes a change in the packet-processing branch that doesn't get copied to the timeout branch.

Based on these tradeoffs, I'm inclined to keep the `biased` approach.

> do we need to move any of the other branches (e.g. shutdown)?

I haven't found any other important races, when both branches are notified, and we need to process one before the other. We might get some performance benefit of processing writes before reads, or the other way around, but I can't see any place where it would cause a problem. Eventually, one branch will always drain out and we'll move onto the next one. It doesn't seem we'll get stuck in any particular branch.

So I generally favored changing the order as little as possible. What's your thinking on moving the shutdown branch?

Reply:
Another potential benefit of keeping `biased` is using it to do something like: process all incoming packets before sending the SYN that ACKs the very latest one.